Create new X509CertificateLoader

bartonjs opened the pull request:
The new certificate loader only loads one data type per method, unlike the previous loader mechanism (new X509Certiicate2(bytes, ...)).  It also allows for caller configuration to control cost-of-work limits and some common usability gotchas around Windows PFX loading.

This change adds the new loader, and changes the X509Certificate2 ctors to use it; a followup will mark the ctors as Obsolete and update usage in the dotnet/runtime codebase.

Contributes to #91763.

bartonjs committed:
Create new X509CertificateLoader

The new certificate loader only loads one data type per method, unlike
the previous loader mechanism (new X509Certiicate2(bytes, ...)).  It also
allows for caller configuration to control cost-of-work limits and some
common usability gotchas around Windows PFX loading.

This change adds the new loader, and changes the X509Certificate2 ctors
to use it; a followup will mark the ctors as Obsolete and update usage in
the dotnet/runtime codebase.

stephentoub commented on src/libraries/Common/src/System/IO/MemoryMappedFiles/MemoryMappedFileMemoryManager.cs:
```diff
@@ -0,0 +1,113 @@
+            _accessor = null!;
+            _mappedFile = null!;
+
+            if (disposing)
+            {
+                GC.SuppressFinalize(this);
+            }
```
Do we need this? MemoryManager's IDisposable.Dispose calls GC.SuppressFinalize.

bartonjs replied:
The best real answer I can come up with is "it probably came from before I realized it could be a MemoryManager", but even that's weird.

So I'll pretend it was intentional and say "clearly, this is the thing I put in so someone would have something to comment about..."

vcsjones commented on src/libraries/Common/src/System/Security/Cryptography/X509Certificates/X509CertificateLoader.cs:
```diff
@@ -0,0 +1,742 @@
+        [DoesNotReturn]
+        private static void ThrowWithHResult(string message, int hResult, Exception innerException)
+        {
+#if NETCOREAPP
```
```suggestion
#if NET
```

stephentoub commented on src/libraries/Common/src/System/Security/Cryptography/X509Certificates/Pkcs12LoaderLimits.cs:
```diff
@@ -0,0 +1,373 @@
+            set
+            {
+                if (value < 0)
+                    throw new ArgumentOutOfRangeException(nameof(value), SR.ArgumentOutOfRange_NeedNonNegNum);
```
Earlier an ifdef was used to use ObjectDisposedException.ThrowIf if it's available. I assume we're not doing so here with AOORE just because there are a bunch of them and it'd be abnoxious?

bartonjs replied:
That was the reason, yeah; but looking at it now in this file it's just repeating a negative check, so making a helper seems reasonable (it'll get re-used)

bartonjs replied:
How "amusing"... I updated every instance except the one this comment was on.  Fixed locally now.

vcsjones commented on src/libraries/Microsoft.Bcl.Cryptography/ref/Microsoft.Bcl.Cryptography.cs:
```diff
@@ -25,3 +26,50 @@ public sealed partial class SP800108HmacCounterKdf : System.IDisposable
+    {
+        public Pkcs12LoadLimitExceededException(string propertyName) { }
+    }
+#if NETCOREAPP
```
```suggestion
#if NET

vcsjones commented on src/libraries/Microsoft.Bcl.Cryptography/src/System/Security/Cryptography/NetStandardShims.cs:
```diff
@@ -21,6 +35,96 @@ internal static unsafe int GetBytes(this Encoding encoding, ReadOnlySpan<char> s
         }
+
+        internal static void ReadExactly(this System.IO.Stream stream, Span<byte> buffer) =>
+            ReadAtLeast(stream, buffer, buffer.Length, throwOnEndOfStream: true);
```
Is delegating this to `ReadAtLeast` correct? If `buffer` is two bytes in length, you are saying it is okay to read three bytes, at which the stream has probably read more than you are accounting for and you threw away one byte.

bartonjs replied:
`ReadAtLeast` can't read more than the size of the buffer, so it's really `ReadRange(minimumBytes, buffer.Length)`.  If `minimumBytes == buffer.Length` the upper and lower bounds are the same, so it's the same as `ReadExactly`.

vcsjones commented on src/libraries/Microsoft.Bcl.Cryptography/src/System/Security/Cryptography/NetStandardShims.cs:
```diff
@@ -21,6 +35,96 @@ internal static unsafe int GetBytes(this Encoding encoding, ReadOnlySpan<char> s
+            }
+        }
+
+#if NETSTANDARD2_0
```
Why does this need a compiler def if the whole file is only for downlevel targets?

bartonjs replied:
Were there a ns2.1 build it wouldn't be needed... and in the netfx build it's not needed, as the entirety of the IncrementalHash class is provided for compat and already includes the function.

But the `#if` does make things uglier, so I've gone ahead and removed it.

vcsjones commented on src/libraries/Microsoft.Bcl.Cryptography/src/System/Security/Cryptography/PbeEncryptionAlgorithm.netstandard.cs:
```diff
@@ -0,0 +1,16 @@
+
+namespace System.Security.Cryptography
+{
+#if !NETCOREAPP
```
```suggestion
#if !NET
```

vcsjones commented on src/libraries/Microsoft.Bcl.Cryptography/src/System/Security/Cryptography/PbeParameters.netstandard.cs:
```diff
@@ -0,0 +1,35 @@
+
+namespace System.Security.Cryptography
+{
+#if !NETCOREAPP
```
```suggestion
#if !NET
```

bartonjs commented on src/libraries/Common/src/System/Security/Cryptography/X509Certificates/X509CertificateLoader.Pkcs12.cs:
```diff
@@ -0,0 +1,944 @@
+                    {
+                        if (!pfxAsn.VerifyMac(password, authSafeContents))
+                        {
+                            password = password.ContainsNull() ? "".AsSpan() : default;
```
Yep.  From the API perspective null and empty are the same thing, but from the algorithm perspective they're different.  So if we got null, try again with empty, and if we got empty, try again with null.

vcsjones commented on src/libraries/Microsoft.Bcl.Cryptography/tests/X509Certificates/TestData.cs:
Is this file a duplicate of `src/libraries/System.Security.Cryptography/tests/X509Certificates/TestData.cs`? Should it be moved under common instead, or use partials, or whatever to de-dupe?

bartonjs replied:
Everything in it is duplicate of TestData, but it's less than a full dupe.  But, yeah, moving the current one to Common probably makes sense.  (And, if it doesn't because of some other content in it, moving this slice as a partial)

vcsjones reviewed (commented):
What I have so far. Scrollbar says I have a ways to go still.

bartonjs commented on src/libraries/Common/src/System/Security/Cryptography/PasswordBasedEncryption.cs:
```diff
@@ -1113,5 +1108,62 @@ private static RC2 CreateRC2()
+                    iterationCount,
+                    prf);
+            }
+            catch (MissingMethodException e)
```
The odds of someone having a 4.6.2 runtime and using this package feel so low to me as to not warrant the complexity of repeating the PBKDF2 implementation.  We can always add it later if it comes up.

bartonjs commented on src/libraries/Common/src/System/Security/Cryptography/X509Certificates/X509CertificateLoader.Pkcs12.cs:
```diff
@@ -0,0 +1,944 @@
+                    }
+                    else
+                    {
+                        // Should there be an option here to preserve this?
+                        // Ignore this?
```
Yeah, I think is a place where I was previously on the side of "we shouldn't block it", but then realized it's "there's a whole chunk here that we just don't know how to read".  It might be worth making a dedicated message for, but I don't think we should skip it (without an option), otherwise there's the oddity of in some theoretical future version a file throws for being over work limits that previously worked (because it skipped this block).

So, I'm starting with removing the comment, but am open to discussion.

vcsjones replied:
I think removing the comment is fine enough. It's the behavior we have now.

bartonjs commented on src/libraries/Common/src/System/Security/Cryptography/X509Certificates/X509CertificateLoader.Pkcs12.cs:
```diff
@@ -0,0 +1,944 @@
+
+            internal bool LockedPassword => (_passwordState & 1) != 0;
+
+            internal void LockPassword()
```
"We're past the point where we're willing to flip between null and empty", so the good old "Lock out changes" button from War Games.  "Lock" and "Freeze" are the only words that come to mind for me here.

It's different from "confirmed", in that "confirmed" means that the PFX had a MAC and the password worked for it.

vcsjones replied:
What if the MAC and PBKDF2 password are different?

vcsjones replied:
@bartonjs did digging yield anything interesting here?

bartonjs replied:
Digging didn't finish, and got deferred to https://github.com/dotnet/runtime/issues/103929

vcsjones replied:
Okay. As long as we have something somewhere that tracks it.

bartonjs commented on src/libraries/Common/src/System/Security/Cryptography/X509Certificates/X509CertificateLoader.Pkcs12.cs:
```diff
@@ -0,0 +1,944 @@
+                writer.Reset();
+
+                HashAlgorithmName hashAlgorithm = HashAlgorithmName.SHA1;
+                byte[] macKey = new byte[20];
```
Pinned, sure.  Zeroed? No, because it gets overwritten with the MAC output value (added a comment).

bartonjs commented on src/libraries/Common/src/System/Security/Cryptography/X509Certificates/X509CertificateLoader.Pkcs12.cs:
```diff
@@ -0,0 +1,944 @@
+                    //   -- Note: The default is for historical reasons and its use is
+                    //   -- deprecated.
+                    // }
+                    writer.PushSequence();
```
I think this code (before moving it to here) is what inspired changing PushSequence from void to returning the state.  Went ahead and changed it.

vcsjones commented on src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/X509Certificates/X509CertificateLoader.Windows.cs:
```diff
@@ -0,0 +1,336 @@
+            }
+            finally
+            {
+                CryptographicOperations.ZeroMemory(MemoryMarshal.AsBytes(szPassword));
```
Probably doesn't make much of a difference but you could slice `szPassword` to `password.Length` to avoid zeroing unwritten portions of the eagerly allocated stack buffer.

vcsjones commented on src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/X509Certificates/X509CertificateLoader.Windows.cs:
```diff
@@ -0,0 +1,336 @@
+                {
+                    if (password.Length >= MaxStackPasswordLength)
+                    {
+                        alloc = NativeMemory.Alloc((uint)password.Length + 1, sizeof(char));
```
This is "fine" but I am wondering if this is an approach that you prefer now. Typically, we would just allocate on the heap, if the password is too big for the stack and put the whole thing in a `fixed`. Something like


```C#
Span<char> szPassword = stackalloc char[MaxStackPasswordLength + 1];

if (password.Length > MaxStackPasswordLength)
{
    szPassword = new char[password.Length + 1];
}

fixed (char* pPassword = szPassword)
{
    password.CopyTo(szPassword);
    szPassword[password.Length] = '\0';

    // Do the thing
}
```

What do we gain using `NativeMemory`, and, is this now "preferred" than using the managed heap for situations like this?

vcsjones replied:
My thoughts on when it is best to use `NativeMemory` is when we need a fixed buffer to survive outside the life of a method call (like the life of a class instance) but that doesn't seem to be the case here.

vcsjones commented on src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/X509Certificates/X509CertificateLoader.iOS.cs:
```diff
@@ -0,0 +1,123 @@
+            }
+
+            if (X509Certificate2.GetCertContentType(data) != X509ContentType.Cert)
+            {
+
+                throw new CryptographicException(SR.Cryptography_Der_Invalid_Encoding);
```
Empty line.
```suggestion
            {
                throw new CryptographicException(SR.Cryptography_Der_Invalid_Encoding);
```

vcsjones commented on src/libraries/System.Security.Cryptography/tests/X509Certificates/CertLoaderTests.cs:
```diff
@@ -0,0 +1,139 @@
+        {
+            protected override X509Certificate2 LoadCertificate(byte[] data)
+            {
+                string path = Path.GetTempFileName();
+
+                try
+                {
+                    File.WriteAllBytes(path, data);
+                    return X509CertificateLoader.LoadCertificateFromFile(path);
+                }
+                finally
+                {
+                    File.Delete(path);
+                }
```
```suggestion
                using (TempFileHolder holder = new(data))
                {
                    return X509CertificateLoader.LoadCertificateFromFile(holder.FilePath);
                }
```

vcsjones commented on src/libraries/System.Security.Cryptography/tests/X509Certificates/CertLoaderTests.cs:
```diff
@@ -0,0 +1,139 @@
+        {
+            protected override X509Certificate2 LoadCertificate(byte[] data)
+            {
+                string path = Path.GetTempFileName();
```
Same `TempFileHolder` suggestion.

vcsjones commented on src/libraries/System.Security.Cryptography/tests/X509Certificates/CertLoaderTests.cs:
```diff
@@ -0,0 +1,139 @@
+            }
+
+            [Fact]
+            [ActiveIssue("Fails as NewFile, NewSpan, NewArray, LegacyArray", TestPlatforms.Linux)]
```
Are these `ActiveIssue` tests things that need to be addressed? If they will be addressed after this pull request, can we write an issue describing the problem and add a link back here?

Same feedback for the rest of the file.

vcsjones commented on src/libraries/System.Security.Cryptography/tests/X509Certificates/CollectionImportTests.cs:
```diff
@@ -30,6 +30,19 @@ public static void ImportEmpty_Pkcs12()
         }
 
+        [Fact]
+        public static void ImportEmpty_Pkcs12_NL()
```
`NL`? Niederlande? New Loader? Is there a better name?

vcsjones commented on src/libraries/System.Security.Cryptography/tests/X509Certificates/CtorTests.cs:
```diff
@@ -198,6 +208,33 @@ public static void TestCopyConstructor_Lifetime_Independent()
+        [Fact]
+        public static void Loader_Lifetime_Independent()
+        {
+            var c1 = X509CertificateLoader.LoadPkcs12(TestData.PfxData, TestData.PfxDataPassword);
+            using (var c2 = X509CertificateLoader.LoadPkcs12(TestData.PfxData, TestData.PfxDataPassword))
```
These aren't typical uses of `var`.

bartonjs commented on src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/X509Certificates/StorePal.macOS.cs:
```diff
@@ -37,22 +37,21 @@ private static ILoaderPal FromBlob(ReadOnlySpan<byte> rawData, SafePasswordHandl
+                try
                 {
-                    throw new PlatformNotSupportedException(SR.Cryptography_X509_NoEphemeralPfx);
+                    return new CollectionBasedLoader(
+                        X509CertificateLoader.LoadPkcs12Collection(
+                            rawData,
+                            password.DangerousGetSpan(),
+                            keyStorageFlags,
+                            X509Certificate.GetPkcs12Limits(readingFromFile, password)));
+                }
+                catch (Pkcs12LoadLimitExceededException e)
+                {
+                    throw new CryptographicException(
+                        SR.Cryptography_X509_PfxWithoutPassword_MaxAllowedIterationsExceeded,
+                        e);
```
This is the compat path for funneling `new X509Certificate2(pfxBytes)` into the new loader.

vcsjones replied:
Sure; that much I understand. I just see this compat path in all of the StorePal implementations. I was idly thinking if it should be "above" the StorePal, but we don't have to solve that here (and maybe it wouldn't work out anyway)

vcsjones commented on src/libraries/Common/tests/System/Security/Cryptography/X509Certificates/X509CertificateLoaderTests.cs:
```diff
@@ -0,0 +1,284 @@
+
+        protected override bool TryGetContentType(byte[] bytes, string path, out X509ContentType contentType)
+        {
+            contentType = X509ContentType.Unknown;
```
Why does the span implementation always return unknown while the array one does a live class to `GetCertContentType`?

bartonjs replied:
All of the data in the tests ultimately comes from a file, or a byte[] (or each independently), so the span version is either seeing redundant data from byte[] or has a Memory-Mapped view of the file.

Combined with GetCertContentType not being available on all TFMs (which matters for the backcompat package) it seemed not worth the hassle, given that this is only ever used as a precondition.

(To which you reply: "Cool.  Add a comment.")

vcsjones commented on src/libraries/Common/tests/System/Security/Cryptography/X509Certificates/X509CertificateLoaderTests.cs:
```diff
@@ -0,0 +1,284 @@
+
+        protected override X509Certificate2 LoadCertificateNoFile(byte[] bytes)
+        {
+            string path = Path.GetTempFileName();
```
Consider using the `TempFileHolder` as recommended somewhere else.

vcsjones commented on src/libraries/Common/tests/System/Security/Cryptography/X509Certificates/X509CertificateLoaderTests.cs:
```diff
@@ -0,0 +1,284 @@
+
+        internal static void AssertRawDataEquals(byte[] expected, X509Certificate2 cert)
+        {
+#if NETCOREAPP
```
```suggestion
#if NET
```

bartonjs committed:
Apply mechanical review feedback

bartonjs commented on src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/X509Certificates/X509Certificate.LegacyLimits.cs:
```diff
@@ -0,0 +1,55 @@
+
+        private static Pkcs12LoaderLimits MakeLegacyLimits()
+        {
+            Pkcs12LoaderLimits limits = new Pkcs12LoaderLimits
```
I think I thought about that at the time, and then expanded it out.  I can't imagine any limit we would add in the future that we wouldn't say this should just inherit the dangerous ones... so... I'll switch it.

bartonjs committed:
Respond to further feedback

bartonjs committed:
Merge remote-tracking branch 'dotnet/main' into x509loader

bartonjs committed:
Finish moving/unifying the TestData file

bartonjs committed:
Add more DisabledOnBrowser attributes

bartonjs committed:
Apply further feedback

vcsjones commented on src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/X509Certificates/X509CertificateLoader.macOS.cs:
```diff
@@ -0,0 +1,230 @@
+                }
+            }
+
+            if (key is DSAImplementation.DSASecurityTransforms dsa)
+            {
+                // DSA always uses legacy CSSM keys do no need to convert
+                return dsa.GetKeys().PrivateKey;
+            }
```
This is, I believe, one of the reasons you are hitting an assert in the unit tests for DSA keys. For RSA and ECDSA, we export and import the keys, which gives us new `SafeSecKeyRefHandle`s with an independent lifetime from the security transform.

That does not appear to be the case with DSA. That is returning the same key handle from the transform. Later on the security transform gets disposed, and the handle gets disposed out from under it.

bartonjs replied:
I'm about to make a push of other things I've cleaned up, and this won't be fixed yet; because I'm doing the push just to get up to date on my mac where I can verify the change.

vcsjones commented on src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/X509Certificates/X509CertificateLoader.Unix.cs:
```diff
@@ -0,0 +1,604 @@
+                            {
+                                // No need to check bytesRead since it was verified
+                                // in PrivateKeyInfoAsn.Decode
+                                key.ImportPkcs8PrivateKey(safeBag.BagValue.Span, out int _);
```
On macOS, this will throw a PNSE for unsupported EC curves (like brainpool) and some tests are expecting a CryptographicException to be thrown. Maybe something like this

```suggestion
                                try
                                {
                                    key.ImportPkcs8PrivateKey(safeBag.BagValue.Span, out int _);
                                }
                                catch (PlatformNotSupportedException pe)
                                {
                                    throw new CryptographicException(SR.Cryptography_NotValidPrivateKey, pe);
                                }
```

vcsjones replied:
You will need the same fix on line 393.

bartonjs replied:
I'm not sure that the second instance is important (since it should have already hit the first) one; but I went ahead and made a helper function and routed them both through it.

bartonjs committed:
Complete the move to TempFileHolder

bartonjs committed:
Fix handling of DSA keys in cert loader on macOS

bartonjs committed:
Clean up test scaffolding from feature standup

bartonjs committed:
Merge remote-tracking branch 'dotnet/main' into x509loader

bartonjs committed:
Fix tests on Linux

bartonjs committed:
Fix test build for netfx

bartonjs committed:
Turn ActiveIssue (skip) into expected failures

bartonjs committed:
Use more varied subject names in WindowsAttribute tests

bartonjs committed:
Fix netfx test build

bartonjs committed:
Revert "Fix netfx test build"

bartonjs committed:
Fix netfx test build

bartonjs committed:
Full swap is important in both Debug and Release

For one, the code as written lost the attributes it should have preserved.
For two, the swapped in attribute might be important.

vcsjones commented:
/azp run runtime-androidemulator

bartonjs committed:
Prevent Android PKCS7 content sniffing in X509 load

bartonjs commented:
/azp run runtime-androidemulator

bartonjs committed:
Use lax reading in Android content filter

bartonjs commented:
/azp run runtime-androidemulator

pavelsavara commented:
this is preventing code flow into azdo internal repo because the security scanner detects credentials.

```
 ! [remote rejected]         main -> main (VS403654: The push was rejected because it contains one or more secrets.

Resolve the following secrets before pushing again. For help, see https://aka.ms/1ESSecretScanning.

Secrets:

commit: d4fe4a7e987584bfa4904dd1027a873029b2005d
paths:
/src/libraries/Common/tests/System/Security/Cryptography/X509Certificates/TestData.cs(2710,1-26) : SEC101/013 : PemPrivateKey
/src/libraries/Common/tests/System/Security/Cryptography/X509Certificates/TestData.cs(2723,1-26) : SEC101/013 : PemPrivateKey
/src/libraries/Common/tests/System/Security/Cryptography/X509Certificates/TestData.cs(2745,1-29) : SEC101/013 : PemPrivateKey
/src/libraries/Common/tests/System/Security/Cryptography/X509Certificates/TestData.cs(2752,1-30) : SEC101/013 : PemPrivateKey
/src/libraries/Common/tests/System/Security/Cryptography/X509Certificates/TestData.cs(2802,1-26) : SEC101/013 : PemPrivateKey
/src/libraries/Common/tests/System/Security/Cryptography/X509Certificates/TestData.cs(2881,1-30) : SEC101/013 : PemPrivateKey
/src/libraries/Common/tests/System/Security/Cryptography/X509Certificates/TestData.cs(2964,1-26) : SEC101/013 : PemPrivateKey
/src/libraries/Common/tests/System/Security/Cryptography/X509Certificates/TestData.cs(57,149-13) : SEC101/055 : Pkcs12PrivateKey
/src/libraries/Common/tests/System/Security/Cryptography/X509Certificates/TestData.cs(1476,41-55) : SEC101/055 : Pkcs12PrivateKey
/src/libraries/Common/tests/System/Security/Cryptography/X509Certificates/TestData.cs(3443,13-26) : SEC101/055 : Pkcs12PrivateKey)
```

bartonjs commented:
Breaking change docs written.  https://github.com/dotnet/docs/issues/42613
